File: src/back-end/FunChess/src/FunChess.Core.Chess/Pieces/Queen.cs
using FunChess.Core.Chess.Enums;
using FunChess.Core.Chess.Extensions;
using FunChess.Core.Chess.Structs;

namespace FunChess.Core.Chess.Pieces;

internal sealed class Queen : Piece
{
    public override bool Move(Board board, Move move)
    {
        if (!MoveIsValid(board, move, out _)) return false;

        board.InternalBoard[move.Next.Index] = board.InternalBoard[move.Previous.Index];
        board.InternalBoard[move.Previous.Index] = Cell.Empty;

        return true;
    }
    
    public override bool MoveIsValid(Board board, Move move, out SpecialMove specialMove)
    {
        specialMove = SpecialMove.None;
        if (move.DiffX == move.DiffY)
        {
            if (Math.Abs(move.DiffX) <= 1)
            {
                return true;
            }
            return move.DiffX < 0 ? board.InternalBoard.CheckSouthwest(move) : board.InternalBoard.CheckNortheast(move);
        }
        if (move.DiffX + move.DiffY == 0)
        {
            if (Math.Abs(move.DiffX) <= 1)
            {
                return true;
            }
            return move.DiffX < 0 ? board.InternalBoard.CheckNorthwest(move) : board.InternalBoard.CheckSoutheast(move);
        }
        if (move.DiffX == 0)
        {
            if (Math.Abs(move.DiffY) <= 1)
            {
                return true;
            }
            return move.DiffY < 0 ? board.InternalBoard.CheckSouth(move) : board.InternalBoard.CheckNorth(move);
        }
        if (move.DiffY != 0) return false;
        if (Math.Abs(move.DiffX) <= 1) return true;
        
        return move.DiffX < 0 ? board.InternalBoard.CheckWest(move) : board.InternalBoard.CheckEast(move);
    }
}

File: src/front-end/fun-chess-app/src/core/chess/pieces/Queen.ts
import Piece from "../Piece";
import Board from "../Board";
import SpecialMove from "../enums/SpecialMove";
import Move from "../structs/Move";
import { checkEast, checkNorth, checkNortheast, checkNorthwest, checkSouth, checkSoutheast, checkSouthwest, checkWest } from "../utils/cells-utils";

class Queen extends Piece {
    private constructor() {
        super(8);
    }

    public static readonly instance: Piece = new Queen();

    public override moveIsValid(board: Board, move: Move): [boolean, SpecialMove] {
        const specialMove = SpecialMove.None;
        if (move.diffX === move.diffY) {
            if (Math.abs(move.diffX) <= 1) {
                return [true, specialMove];
            }
            return [move.diffX < 0 ? checkSouthwest(board.internalBoard, move) : checkNortheast(board.internalBoard, move), specialMove];
        }
        if (move.diffX + move.diffY === 0) {
            if (Math.abs(move.diffX) <= 1) {
                return [true, specialMove];
            }
            return [move.diffX < 0 ? checkNorthwest(board.internalBoard, move) : checkSoutheast(board.internalBoard, move), specialMove];
        }
        if (move.diffX === 0) {
            if (Math.abs(move.diffY) <= 1) {
                return [true, specialMove];
            }
            return [move.diffY < 0 ? checkSouth(board.internalBoard, move) : checkNorth(board.internalBoard, move), specialMove];
        }
        if (move.diffY !== 0) return [false, specialMove];
        if (Math.abs(move.diffX) <= 1) return [true, specialMove];
        
        return [move.diffX < 0 ? checkWest(board.internalBoard, move) : checkEast(board.internalBoard, move), specialMove];
    }
}

export default Queen;